crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -36,68 +36,122 @@ impl FormatNodeRule<ExprBinOp> for FormatExprBinOp {
+                let needs_space = !is_simple_power_expression(item);
+
+                let before_operator_space = if needs_space {
+                    soft_line_break_or_space()
+                } else {
+                    soft_line_break()
+                };
+
+                write!(
+                    f,
+                    [
+                        left.format(),
+                        before_operator_space,
+                        op.format(),
+                        trailing_comments(operator_comments),
+                    ]
+                )?;
+
+                // Format the operator on its own line if the right side has any leading comments.
+                if comments.has_leading_comments(right.as_ref()) {
+                    write!(f, [hard_line_break()])?;
+                } else if needs_space {
+                    write!(f, [space()])?;
+                }
+
+                write!(f, [group(&right.format())])
             }
 
-            write!(f, [group(&right.format())])
+            BinaryLayout::ExpandLeft => {
+                let left = left.format().memoized();
+                let right = right.format().memoized();
+                write!(
+                    f,
+                    [best_fitting![
+                        // Everything on a single line
+                        format_args![left, space(), op.format(), space(), right],
+                        // Break the left over multiple lines, keep the right flat
+                        format_args![
+                            group(&left).should_expand(true),
+                            space(),
+                            op.format(),
+                            space(),
+                            right
+                        ],
+                        // The content doesn't fit, indent the content and break after the operator.
+                        format_args![
+                            text("("),
+                            block_indent(&format_args![
+                                left,
+                                hard_line_break(),
+                                op.format(),
+                                space(),
+                                right
+                            ]),
+                            text(")")
+                        ]
+                    ]
+                    .with_mode(BestFittingMode::AllLines)]
+                )
+            }
+
+            BinaryLayout::ExpandRight => {
+                let left_group = f.group_id("BinaryLeft");
+
+                write!(
+                    f,
+                    [
+                        // Wrap the left in a group and gives it an id. The printer first breaks the
+                        // right side if `right` contains any line break because the printer breaks
+                        // sequences of groups from right to left.
+                        // Indents the left side if the group breaks.
```

Comment:
i'm having trouble following this code, shouldn't right breaks and if right is expanded but this is insufficient the whole thing should be parenthesized and indented?

Reply:
I'm not entirely sure if I understand this comment. 

The formatter measures the left group to see if it fits. It does that by measuring if all the content *up to* the first line break (soft or hard) in the right group (or whatever comes after) fits on a line. It doesn't expand the left group if that's the case. 

This is is the behavior that we want. We want to expand the right first and not add parentheses in that case. We only want to add parentheses and indent the code if the left expands. That's why `indent_if_group_breaks` and the `if_group_breaks` refer to the left group and not their enclosing group.